test/solver/solver.cpp

```diff
@@ -0,0 +1,742 @@
+        auto result = VecType::create(ref, size, stride);
+        result->read(gen_dense_data<typename VecType::value_type,
+                                    typename Mtx::index_type>(size));
+        return {std::move(result), exec};
+    }
+
+    template <typename VecType>
+    double tol(const test_pair<VecType>& x)
+    {
+        return Config::tolerance() * std::sqrt(x.ref->get_size()[1]);
+    }
+
+    template <typename VecType>
+    double mixed_tol(const test_pair<VecType>& x)
+    {
+        return std::max(r_mixed<value_type, mixed_value_type>() *
+                            std::sqrt(x.ref->get_size()[1]),
+                        tol(x));
+    }
+
+    template <typename TestFunction>
+    void forall_matrix_scenarios(TestFunction fn)
+    {
+        auto guarded_fn = [&](auto mtx) {
+            try {
+                fn(std::move(mtx));
+            } catch (std::exception& e) {
+                FAIL() << e.what();
+            }
+        };
+        {
+            SCOPED_TRACE("Empty matrix (0x0)");
+            guarded_fn(gen_mtx(0, 0, 0, 0));
+        }
+        {
+            SCOPED_TRACE("Sparse Matrix with variable row nnz (50x50)");
+            guarded_fn(gen_mtx(50, 50, 10, 50));
+        }
+    }
+
+    template <typename TestFunction>
+    void forall_solver_scenarios(const test_pair<Mtx>& mtx, TestFunction fn)
+    {
+        auto guarded_fn = [&](auto solver) {
+            try {
+                fn(std::move(solver));
+            } catch (std::exception& e) {
+                FAIL() << e.what();
+            }
+        };
+        {
+            SCOPED_TRACE("Unpreconditioned solver with 0 iterations");
+            guarded_fn(test_pair<SolverType>{
+                Config::build(ref, 0).on(ref)->generate(mtx.ref),
+                Config::build(exec, 0).on(exec)->generate(mtx.dev)});
+        }
+        if (Config::is_preconditionable()) {
+            SCOPED_TRACE("Preconditioned solver with 0 iterations");
+            guarded_fn(test_pair<SolverType>{
+                Config::build_preconditioned(ref, 0).on(ref)->generate(mtx.ref),
+                Config::build_preconditioned(exec, 0).on(exec)->generate(
+                    mtx.dev)});
+        }
+        if (Config::is_iterative()) {
+            {
+                SCOPED_TRACE("Unpreconditioned solver with 4 iterations");
+                guarded_fn(test_pair<SolverType>{
+                    Config::build(ref, 4).on(ref)->generate(mtx.ref),
+                    Config::build(exec, 4).on(exec)->generate(mtx.dev)});
+            }
+            if (Config::is_preconditionable()) {
+                SCOPED_TRACE("Preconditioned solver with 4 iterations");
+                guarded_fn(test_pair<SolverType>{
+                    Config::build_preconditioned(ref, 4).on(ref)->generate(
+                        mtx.ref),
+                    Config::build_preconditioned(exec, 4).on(exec)->generate(
+                        mtx.dev)});
+            }
+        }
```

Comment:
You could also check for a `is_restartable` and test with 0 restart, 1 restart and some normal restart count. This could also include subspace dimension for IDR.

Reply:
I tried incorporating this by using two GMRES configurations: One that restarts when doing 4 iterations, and one that doesn't. How would you behave differently if `is_restartable` is available or not?

Reply:
Indeed, ideally, there should be no difference. But it would be nice just as a check for how the restart parameter is treated, since it and the max iterations are two separate variables in the implementation.